File: funnyvote/app/src/test/java/com/heaton/funnyvote/ui/votedetail/VoteDetailPresenterTest.java
package com.heaton.funnyvote.ui.votedetail;

import com.heaton.funnyvote.data.VoteData.VoteDataRepository;
import com.heaton.funnyvote.data.VoteData.VoteDataSource;
import com.heaton.funnyvote.data.user.UserDataRepository;
import com.heaton.funnyvote.data.user.UserDataSource;
import com.heaton.funnyvote.database.Option;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.database.VoteData;

import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyBoolean;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.anyList;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Unit tests for the implementation of {@link VoteDetailPresenter}
 */
public class VoteDetailPresenterTest {
    private static User user;
    private static final String voteCode = "CODE_123";
    private static final VoteData voteData = new VoteData();
    private static List<Option> optionList = new ArrayList<>();

    @Mock
    private UserDataRepository userDataRepository;
    @Mock
    private VoteDataRepository voteDataRepository;
    @Mock
    private VoteDetailContract.View view;

    private VoteDetailPresenter presenter;
    @Captor
    private ArgumentCaptor<UserDataSource.GetUserCallback> getUserCallbackArgumentCaptor;
    @Captor
    private ArgumentCaptor<VoteDataSource.GetVoteDataCallback> getVoteDataCallbackArgumentCaptor;
    @Captor
    private ArgumentCaptor<VoteDataSource.GetVoteOptionsCallback> getVoteOptionsCallbackArgumentCaptor;
    @Captor
    private ArgumentCaptor<VoteDataSource.FavoriteVoteCallback> favoriteVoteCallbackArgumentCaptor;
    @Captor
    private ArgumentCaptor<VoteDataSource.PollVoteCallback> pollVoteCallbackArgumentCaptor;
    @Captor
    private ArgumentCaptor<VoteDataSource.AddNewOptionCallback> addNewOptionCallbackArgumentCaptor;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);

        user = mock(User.class);
        when(user.getUserName()).thenReturn("Heaton");
    }

    @Test
    public void createPresenter_setsThePresenterToView() {
        // Get a reference to the class under test
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);

        // Then the presenter is set to the view
        verify(view).setPresenter(presenter);
    }

    @Test
    public void getVoteDataFromRepositoryAndLoadIntoView() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.start();
        verify(userDataRepository).getUser(getUserCallbackArgumentCaptor.capture(), eq(false));
        InOrder inOrder = Mockito.inOrder(view);
        inOrder.verify(view).showLoadingCircle();
        getUserCallbackArgumentCaptor.getValue().onResponse(user);
        verify(view).setUpAdMob(eq(user));
        verify(voteDataRepository).getVoteData(eq(voteCode), eq(user), getVoteDataCallbackArgumentCaptor.capture());
        getVoteDataCallbackArgumentCaptor.getValue().onVoteDataLoaded(voteData);
        inOrder.verify(view).hideLoadingCircle();
        verify(voteDataRepository).getOptions(eq(voteData), getVoteOptionsCallbackArgumentCaptor.capture());
        getVoteOptionsCallbackArgumentCaptor.getValue().onVoteOptionsLoaded(optionList);
        verify(view).setUpOptionAdapter(eq(voteData), any(Integer.class), eq(optionList));
    }

    @Test
    public void searchOptionAndUpdateToView() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.searchOption("test");
        verify(view).updateSearchView(any(List.class), eq(true));
    }

    @Test
    public void favoriteVoteAndUpdateToView() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.favoriteVote();
        verify(voteDataRepository).favoriteVote(anyString(), anyBoolean(), any(User.class)
                , favoriteVoteCallbackArgumentCaptor.capture());
        favoriteVoteCallbackArgumentCaptor.getValue().onSuccess(voteData.getIsFavorite());
        verify(view).updateFavoriteView(anyBoolean());
        verify(view).showHintToast(anyInt());
    }

    @Test
    public void pollVoteAndUpdateToView() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.pollVote("password");
        InOrder inOrder = Mockito.inOrder(view);
        verify(voteDataRepository).pollVote(anyString(), eq("password"), anyList(), any(User.class)
                , pollVoteCallbackArgumentCaptor.capture());
        inOrder.verify(view).showLoadingCircle();
        pollVoteCallbackArgumentCaptor.getValue().onSuccess(voteData);
        inOrder.verify(view).hideLoadingCircle();
        verify(view).setUpViews(any(VoteData.class), anyInt());
        verify(view).setUpOptionAdapter(any(VoteData.class), anyInt(), anyList());
        verify(view).setUpSubmit(anyInt());
        verify(view).refreshOptions();
        verify(view).hidePollPasswordDialog();
    }

    @Test
    public void pollVoteNeedPWAndShakeDialogAfterShowPWDialog() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.getVoteData().setIsNeedPassword(true);
        presenter.pollVote("password");
        InOrder inOrder = Mockito.inOrder(view);
        inOrder.verify(view).showPollPasswordDialog();
        when(view.isPasswordDialogShowing()).thenReturn(true);
        presenter.pollVote("password");
        verify(voteDataRepository).pollVote(anyString(), eq("password"), anyList(), any(User.class)
                , pollVoteCallbackArgumentCaptor.capture());
        inOrder.verify(view).showLoadingCircle();
        pollVoteCallbackArgumentCaptor.getValue().onPasswordInvalid();
        inOrder.verify(view).shakePollPasswordDialog();
        inOrder.verify(view).hideLoadingCircle();
    }

    @Test
    public void pollVoteFailureAndShowPWDialog() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.getVoteData().setIsNeedPassword(false);
        presenter.pollVote("password");
        InOrder inOrder = Mockito.inOrder(view);
        verify(voteDataRepository).pollVote(anyString(), eq("password"), anyList(), any(User.class)
                , pollVoteCallbackArgumentCaptor.capture());
        inOrder.verify(view).showLoadingCircle();
        pollVoteCallbackArgumentCaptor.getValue().onFailure();
        inOrder.verify(view).hideLoadingCircle();
    }

    @Test
    public void addNewOptionNeedPWAndShakeDialogAfterShowPWDialog() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.getVoteData().setIsNeedPassword(true);
        presenter.addNewOptionCompleted("password", "newOptionText");
        InOrder inOrder = Mockito.inOrder(view);
        inOrder.verify(view).showAddNewOptionPasswordDialog(anyString());
        when(view.isPasswordDialogShowing()).thenReturn(true);
        presenter.addNewOptionCompleted("password", "newOptionText");
        verify(voteDataRepository).addNewOption(anyString(), eq("password"), anyList(), any(User.class)
                , addNewOptionCallbackArgumentCaptor.capture());
        inOrder.verify(view).showLoadingCircle();
        addNewOptionCallbackArgumentCaptor.getValue().onPasswordInvalid();
        inOrder.verify(view).shakeAddNewOptionPasswordDialog();
        inOrder.verify(view).hideLoadingCircle();
    }

    @Test
    public void addNewOptionFailureAndShowPWDialog() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.getVoteData().setIsNeedPassword(false);
        presenter.addNewOptionCompleted("password", "newOptionText");
        InOrder inOrder = Mockito.inOrder(view);
        verify(voteDataRepository).addNewOption(anyString(), eq("password"), anyList(), any(User.class)
                , addNewOptionCallbackArgumentCaptor.capture());
        inOrder.verify(view).showLoadingCircle();
        addNewOptionCallbackArgumentCaptor.getValue().onFailure();
        inOrder.verify(view).hideLoadingCircle();
    }

    @Test
    public void clickOnOption_refreshOptionChoiceTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.resetOptionChoiceStatus(1, "OPTION_123");
        verify(view).updateChoiceOptions(anyList());
    }

    @Test
    public void clickOnOption_refreshOptionExpandTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.resetOptionExpandStatus("OPTION_123");
        verify(view).updateExpandOptions(anyList());
    }

    @Test
    public void clickAddNewOption_addNewOptionStartTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.addNewOptionStart();
        verify(view).refreshOptions();
    }

    @Test
    public void edit_addNewOptionContentReviseTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.addNewOptionContentRevise(1, "inputText");

    }

    @Test
    public void clickComplete_addNewOptionCompletedAndUpdateToView() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.addNewOptionCompleted("password", "newOptionText");
        InOrder inOrder = Mockito.inOrder(view);
        verify(voteDataRepository).addNewOption(anyString(), eq("password"), anyList(), any(User.class)
                , addNewOptionCallbackArgumentCaptor.capture());
        inOrder.verify(view).showLoadingCircle();
        addNewOptionCallbackArgumentCaptor.getValue().onSuccess(voteData);
        inOrder.verify(view).hideLoadingCircle();
        verify(view).setUpViews(any(VoteData.class), anyInt());
        verify(view).setUpOptionAdapter(any(VoteData.class), anyInt(), anyList());
        verify(view).setUpSubmit(anyInt());
        verify(view).refreshOptions();
        verify(view).hideAddNewOptionPasswordDialog();
    }

    @Test
    public void clickOnOption_removeOptionTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.removeOption(1);
        verify(view).refreshOptions();
    }

    @Test
    public void clickOnFab_changeOptionTypeTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.isUserPreResult = false;
        presenter.changeOptionType();
        verify(view).showResultOption(anyInt());
        presenter.isUserPreResult = true;
        presenter.changeOptionType();
        verify(view).showUnPollOption(anyInt());
    }

    @Test
    public void clickOnFab_CheckSortOptionTypeTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.CheckSortOptionType();
        verify(view).showSortOptionDialog(any(VoteData.class));
    }

    @Test
    public void clickOnFab_sortOptionsTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.sortOptions(0);
        verify(view).refreshOptions();
    }

    @Test
    public void clickOnToolbar_IntentToVoteInfoTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.IntentToVoteInfo();
        verify(view).showVoteInfoDialog(any(VoteData.class));
    }

    @Test
    public void clickOnTitle_IntentToTitleDetailTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.IntentToTitleDetail();
        verify(view).showTitleDetailDialog(any(VoteData.class));
    }

    @Test
    public void clickOnMainBar_IntentToShareDialogTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.IntentToShareDialog();
        verify(view).showShareDialog(any(VoteData.class));
    }

    @Test
    public void clickOnMainBar_IntentToAuthorDetailTest() {
        presenter = new VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view);
        presenter.IntentToAuthorDetail();
        verify(view).showAuthorDetail(any(VoteData.class));
    }
}

File: funnyvote/app/src/test/java/com/heaton/funnyvote/ui/votedetail/VoteDetailPresenterTest.kt
package com.heaton.funnyvote.ui.votedetail

import com.heaton.funnyvote.any
import com.heaton.funnyvote.capture
import com.heaton.funnyvote.data.VoteData.VoteDataRepository
import com.heaton.funnyvote.data.VoteData.VoteDataSource
import com.heaton.funnyvote.data.user.UserDataRepository
import com.heaton.funnyvote.data.user.UserDataSource
import com.heaton.funnyvote.database.Option
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.database.VoteData
import com.heaton.funnyvote.eq
import com.heaton.funnyvote.ui.votedetail.VoteDetailContract.View
import org.junit.Before
import org.junit.Test
import org.mockito.*
import org.mockito.Mockito.*
import java.util.*

/**
 * Unit tests for the implementation of [VoteDetailPresenter]
 */
class VoteDetailPresenterTest {

    @Mock
    private lateinit var userDataRepository: UserDataRepository
    @Mock
    private lateinit var voteDataRepository: VoteDataRepository
    @Mock
    private lateinit var view: VoteDetailContract.View

    private lateinit var presenter: VoteDetailPresenter
    @Captor
    private lateinit var getUserCallbackArgumentCaptor: ArgumentCaptor<UserDataSource.GetUserCallback>
    @Captor
    private lateinit var getVoteDataCallbackArgumentCaptor: ArgumentCaptor<VoteDataSource.GetVoteDataCallback>
    @Captor
    private lateinit var getVoteOptionsCallbackArgumentCaptor: ArgumentCaptor<VoteDataSource.GetVoteOptionsCallback>
    @Captor
    private lateinit var favoriteVoteCallbackArgumentCaptor: ArgumentCaptor<VoteDataSource.FavoriteVoteCallback>
    @Captor
    private lateinit var pollVoteCallbackArgumentCaptor: ArgumentCaptor<VoteDataSource.PollVoteCallback>
    @Captor
    private lateinit var addNewOptionCallbackArgumentCaptor: ArgumentCaptor<VoteDataSource.AddNewOptionCallback>

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)

        user = mock(User::class.java)
        `when`(user.getUserName()).thenReturn("Heaton")
    }

    @Test
    fun createPresenter_setsThePresenterToView() {
        // Get a reference to the class under test
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)

        // Then the presenter is set to the view
        verify(view).setPresenter(presenter)
    }

    @Test
    fun getVoteDataFromRepositoryAndLoadIntoView() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.start()
        verify<UserDataRepository>(userDataRepository).getUser(capture(getUserCallbackArgumentCaptor), ArgumentMatchers.eq(false))
        val inOrder = Mockito.inOrder(view)
        inOrder.verify<VoteDetailContract.View>(view).showLoadingCircle()
        getUserCallbackArgumentCaptor.value.onResponse(user)
        verify<View>(view).setUpAdMob(eq(user))
        verify<VoteDataRepository>(voteDataRepository).getVoteData(eq(voteCode), eq(user), getVoteDataCallbackArgumentCaptor.capture())
        getVoteDataCallbackArgumentCaptor.value.onVoteDataLoaded(voteData)
        inOrder.verify<View>(view).hideLoadingCircle()
        verify<VoteDataRepository>(voteDataRepository).getOptions(eq(voteData), capture(getVoteOptionsCallbackArgumentCaptor))
        getVoteOptionsCallbackArgumentCaptor.value.onVoteOptionsLoaded(optionList)
        verify<View>(view).setUpOptionAdapter(eq(voteData), ArgumentMatchers.anyInt(), eq(optionList))
    }

    @Test
    fun searchOptionAndUpdateToView() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.searchOption("test")
        verify<View>(view).updateSearchView(ArgumentMatchers.anyList(), ArgumentMatchers.eq(true))
    }

    @Test
    fun favoriteVoteAndUpdateToView() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData.voteCode = "VOTE_CODE_123"
        presenter.favoriteVote()
        verify<VoteDataRepository>(voteDataRepository).favoriteVote(ArgumentMatchers.anyString()
                , ArgumentMatchers.anyBoolean(), any<User>(), capture(favoriteVoteCallbackArgumentCaptor))
        favoriteVoteCallbackArgumentCaptor.value.onSuccess(voteData.isFavorite)
        verify<View>(view).updateFavoriteView(ArgumentMatchers.anyBoolean())
        verify<View>(view).showHintToast(ArgumentMatchers.anyInt())
    }

    @Test
    fun pollVoteAndUpdateToView() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData = voteData
        presenter.pollVote("password")
        val inOrder = Mockito.inOrder(view)
        verify<VoteDataRepository>(voteDataRepository).pollVote(ArgumentMatchers.anyString(), eq("password")
                , ArgumentMatchers.anyList(), any<User>(), pollVoteCallbackArgumentCaptor.capture())
        inOrder.verify<View>(view).showLoadingCircle()
        pollVoteCallbackArgumentCaptor.value.onSuccess(voteData)
        inOrder.verify<View>(view).hideLoadingCircle()
        verify<View>(view).setUpViews(any<VoteData>(), ArgumentMatchers.anyInt())
        verify<View>(view).setUpOptionAdapter(any<VoteData>(), ArgumentMatchers.anyInt(), ArgumentMatchers.anyList())
        verify<View>(view).setUpSubmit(ArgumentMatchers.anyInt())
        verify<View>(view).refreshOptions()
        verify<View>(view).hidePollPasswordDialog()
    }

    @Test
    fun pollVoteNeedPWAndShakeDialogAfterShowPWDialog() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData.isNeedPassword = true
        presenter.voteData.voteCode = "VOTE_CODE_123"
        presenter.pollVote("password")
        val inOrder = Mockito.inOrder(view)
        inOrder.verify<View>(view).showPollPasswordDialog()
        `when`(view.isPasswordDialogShowing).thenReturn(true)
        presenter.pollVote("password")
        verify<VoteDataRepository>(voteDataRepository).pollVote(ArgumentMatchers.anyString(), eq("password"), ArgumentMatchers.anyList(), any<User>(), pollVoteCallbackArgumentCaptor.capture())
        inOrder.verify<View>(view).showLoadingCircle()
        pollVoteCallbackArgumentCaptor.value.onPasswordInvalid()
        inOrder.verify<View>(view).shakePollPasswordDialog()
        inOrder.verify<View>(view).hideLoadingCircle()
    }

    @Test
    fun pollVoteFailureAndShowPWDialog() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData = voteData
        presenter.voteData.isNeedPassword = false
        presenter.pollVote("password")
        val inOrder = Mockito.inOrder(view)
        verify<VoteDataRepository>(voteDataRepository).pollVote(ArgumentMatchers.anyString(), eq("password"), ArgumentMatchers.anyList(), any<User>(), pollVoteCallbackArgumentCaptor.capture())
        inOrder.verify<View>(view).showLoadingCircle()
        pollVoteCallbackArgumentCaptor.value.onFailure()
        inOrder.verify<View>(view).hideLoadingCircle()
    }

    @Test
    fun addNewOptionNeedPWAndShakeDialogAfterShowPWDialog() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData.isNeedPassword = true
        presenter.voteData.voteCode = voteCode
        presenter.addNewOptionCompleted("password", "newOptionText")
        val inOrder = Mockito.inOrder(view)
        inOrder.verify<View>(view).showAddNewOptionPasswordDialog(ArgumentMatchers.anyString())
        `when`(view.isPasswordDialogShowing).thenReturn(true)
        presenter.addNewOptionCompleted("password", "newOptionText")
        verify<VoteDataRepository>(voteDataRepository).addNewOption(ArgumentMatchers.anyString(), eq("password"), ArgumentMatchers.anyList(), any<User>(), capture(addNewOptionCallbackArgumentCaptor))
        inOrder.verify<View>(view).showLoadingCircle()
        addNewOptionCallbackArgumentCaptor.value.onPasswordInvalid()
        inOrder.verify<View>(view).shakeAddNewOptionPasswordDialog()
        inOrder.verify<View>(view).hideLoadingCircle()
    }

    @Test
    fun addNewOptionFailureAndShowPWDialog() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData = voteData
        presenter.voteData.isNeedPassword = false

        presenter.addNewOptionCompleted("password", "newOptionText")
        val inOrder = Mockito.inOrder(view)
        verify<VoteDataRepository>(voteDataRepository).addNewOption(ArgumentMatchers.anyString()
                , eq("password"), ArgumentMatchers.anyList(), any<User>(), capture(addNewOptionCallbackArgumentCaptor))
        inOrder.verify<View>(view).showLoadingCircle()
        addNewOptionCallbackArgumentCaptor.value.onFailure()
        inOrder.verify<View>(view).hideLoadingCircle()
    }

    @Test
    fun clickOnOption_refreshOptionChoiceTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.resetOptionChoiceStatus(1, "OPTION_123")
        verify<View>(view).updateChoiceOptions(ArgumentMatchers.anyList())
    }

    @Test
    fun clickOnOption_refreshOptionExpandTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.resetOptionExpandStatus("OPTION_123")
        verify<View>(view).updateExpandOptions(ArgumentMatchers.anyList())
    }

    @Test
    fun clickAddNewOption_addNewOptionStartTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.addNewOptionStart()
        verify<View>(view).refreshOptions()
    }

    @Test
    fun edit_addNewOptionContentReviseTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData = voteData
        presenter.addNewOptionContentRevise(1, "inputText")

    }

    @Test
    fun clickComplete_addNewOptionCompletedAndUpdateToView() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData = voteData
        presenter.addNewOptionCompleted("password", "newOptionText")
        val inOrder = Mockito.inOrder(view)
        verify<VoteDataRepository>(voteDataRepository).addNewOption(ArgumentMatchers.anyString(), eq("password"), ArgumentMatchers.anyList(), any<User>(), capture(addNewOptionCallbackArgumentCaptor))
        inOrder.verify<View>(view).showLoadingCircle()
        addNewOptionCallbackArgumentCaptor.value.onSuccess(voteData)
        inOrder.verify<View>(view).hideLoadingCircle()
        verify<View>(view).setUpViews(any<VoteData>(), ArgumentMatchers.anyInt())
        verify<View>(view).setUpOptionAdapter(any<VoteData>(), ArgumentMatchers.anyInt(), ArgumentMatchers.anyList())
        verify<View>(view).setUpSubmit(ArgumentMatchers.anyInt())
        verify<View>(view).refreshOptions()
        verify<View>(view).hideAddNewOptionPasswordDialog()
    }

    @Test
    fun clickOnOption_removeOptionTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.removeOption(1)
        verify<View>(view).refreshOptions()
    }

    @Test
    fun clickOnFab_changeOptionTypeTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.isUserPreResult = false
        presenter.changeOptionType()
        verify<View>(view).showResultOption(ArgumentMatchers.anyInt())
        presenter.isUserPreResult = true
        presenter.changeOptionType()
        verify<View>(view).showUnPollOption(ArgumentMatchers.anyInt())
    }

    @Test
    fun clickOnFab_CheckSortOptionTypeTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.CheckSortOptionType()
        verify<View>(view).showSortOptionDialog(any<VoteData>())
    }

    @Test
    fun clickOnFab_sortOptionsTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.voteData.netOptions = optionList
        presenter.voteData.voteCode = voteCode;
        presenter.sortOptions(0)
        verify<View>(view).updateCurrentOptionsOrder(ArgumentMatchers.anyList());
    }

    @Test
    fun clickOnToolbar_IntentToVoteInfoTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.IntentToVoteInfo()
        verify<View>(view).showVoteInfoDialog(any<VoteData>())
    }

    @Test
    fun clickOnTitle_IntentToTitleDetailTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.IntentToTitleDetail()
        verify<View>(view).showTitleDetailDialog(any<VoteData>())
    }

    @Test
    fun clickOnMainBar_IntentToShareDialogTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.IntentToShareDialog()
        verify<View>(view).showShareDialog(any<VoteData>())
    }

    @Test
    fun clickOnMainBar_IntentToAuthorDetailTest() {
        presenter = VoteDetailPresenter(voteCode, voteDataRepository, userDataRepository, view)
        presenter.IntentToAuthorDetail()
        verify<View>(view).showAuthorDetail(any<VoteData>())
    }

    companion object {
        private lateinit var user: User
        private val voteCode = "CODE_123"
        private val optionList = ArrayList<Option>()
        private val voteData = VoteData().apply {
            voteCode = "CODE_123"
            netOptions = optionList
        }
    }
}
